dds/DCPS/DiscoveryBase.h

```diff
@@ -1247,40 +1271,56 @@ namespace OpenDDS {
+        // part in its TypeInformation? If it's possible, then we may need to handle that case, i.e.,
+        // request only the remote complete TypeObject (if it's not already in the cache).
+        // The following code assumes when the "minimal" part is not included in the discovered
+        // endpoint's TypeInformation, then the "complete" part also is not included.
         if ((writer_type_info->minimal.typeid_with_size.type_id.kind() != XTypes::TK_NONE) &&
             (reader_type_info->minimal.typeid_with_size.type_id.kind() != XTypes::TK_NONE)) {
           if (!writer_local && reader_local) {
-            if (type_lookup_service_ &&
-                !type_lookup_service_->type_object_in_cache(
-                  writer_type_info->minimal.typeid_with_size.type_id)) {
+            const bool need_minimal_tobjs = type_lookup_service_ &&
+              !type_lookup_service_->type_object_in_cache(writer_type_info->minimal.typeid_with_size.type_id);
+            const bool need_complete_tobjs = type_lookup_service_ && use_xtypes_complete_ &&
+              writer_type_info->complete.typeid_with_size.type_id.kind() != XTypes::TK_NONE &&
+              !type_lookup_service_->type_object_in_cache(writer_type_info->complete.typeid_with_size.type_id);
+
+            if (need_minimal_tobjs || need_complete_tobjs) {
               if (DCPS_debug_level >= 4) {
                 ACE_DEBUG((LM_DEBUG, "(%P|%t) EndpointManager::match: Remote Writer\n"));
               }
               bool is_discovery_protected = false;
 #ifdef OPENDDS_SECURITY
               is_discovery_protected = lsi->second.security_attribs_.base.is_discovery_protected;
 #endif
-              save_matching_data_and_get_typeobjects(
-                writer_type_info, md, MatchingPair(writer, reader), writer, is_discovery_protected);
+              save_matching_data_and_get_typeobjects(writer_type_info, md,
+                                                     MatchingPair(writer, reader),
+                                                     writer, is_discovery_protected,
+                                                     need_minimal_tobjs, need_complete_tobjs);
               return;
             }
           } else if (!reader_local && writer_local) {
-            if (type_lookup_service_ && !type_lookup_service_->type_object_in_cache(
-                reader_type_info->minimal.typeid_with_size.type_id)) {
+            const bool need_minimal_tobjs = type_lookup_service_ &&
+              !type_lookup_service_->type_object_in_cache(reader_type_info->minimal.typeid_with_size.type_id);
+            const bool need_complete_tobjs = type_lookup_service_ && use_xtypes_complete_ &&
+              reader_type_info->complete.typeid_with_size.type_id.kind() != XTypes::TK_NONE &&
+              !type_lookup_service_->type_object_in_cache(reader_type_info->complete.typeid_with_size.type_id);
+
```

Comment:
Since this is the same for remote-reader as the code above for remote-writer, could it be refactored to a common method?